velox/exec/tests/ContainerRowSerdeTest.cpp

```diff
@@ -0,0 +1,122 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/ContainerRowSerde.h"
+#include <gtest/gtest.h>
+#include "velox/common/memory/HashStringAllocator.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+namespace facebook::velox::exec {
+
+namespace {
+
+class ContainerRowSerdeTest : public testing::Test,
+                              public velox::test::VectorTestBase {
+ protected:
+  HashStringAllocator::Position serialize(const VectorPtr& data) {
+    ByteStream out(&allocator_);
+    auto position = allocator_.newWrite(out);
+    for (auto i = 0; i < data->size(); ++i) {
+      ContainerRowSerde::serialize(*data, i, out);
+    }
+    allocator_.finishWrite(out, 0);
+    return position;
+  }
+
+  VectorPtr deserialize(
+      HashStringAllocator::Position position,
+      const TypePtr& type,
+      vector_size_t numRows) {
+    auto data = BaseVector::create(type, numRows, pool());
+
+    ByteStream in;
+    HashStringAllocator::prepareRead(position.header, in);
+    for (auto i = 0; i < numRows; ++i) {
+      ContainerRowSerde::deserialize(in, i, data.get());
+    }
+    return data;
+  }
+
+  void testRoundTrip(const VectorPtr& data) {
+    auto position = serialize(data);
+    auto copy = deserialize(position, data->type(), data->size());
+    test::assertEqualVectors(data, copy);
+
+    allocator_.clear();
+  }
+
+  HashStringAllocator allocator_{pool()};
+};
+
+TEST_F(ContainerRowSerdeTest, bigint) {
+  auto data = makeFlatVector<int64_t>({1, 2, 3, 4, 5});
+
+  testRoundTrip(data);
+}
+
+TEST_F(ContainerRowSerdeTest, string) {
+  auto data =
+      makeFlatVector<std::string>({"a", "Abc", "Long test sentence.", "", "d"});
+
+  testRoundTrip(data);
+}
+
+TEST_F(ContainerRowSerdeTest, arrayOfBigint) {
+  auto data = makeArrayVector<int64_t>({
+      {1, 2, 3},
+      {4, 5},
+      {6},
+      {},
+  });
+
+  testRoundTrip(data);
+
+  data = makeNullableArrayVector<int64_t>({
+      {{{1, std::nullopt, 2, 3}}},
+      {{{std::nullopt, 4, 5}}},
+      {{{6, std::nullopt}}},
+      {{{}}},
+  });
+
+  testRoundTrip(data);
+}
+
+TEST_F(ContainerRowSerdeTest, arrayOfString) {
+  auto data = makeArrayVector<std::string>({
+      {"a", "b", "Longer string ...."},
+      {"c", "Abc", "Mountains and rivers"},
+      {},
+      {"Oceans and skies"},
+  });
+
+  testRoundTrip(data);
+
+  data = makeNullableArrayVector<std::string>({
+      {{{std::nullopt,
+         std::nullopt,
+         "a",
+         std::nullopt,
+         "b",
+         "Longer string ...."}}},
+      {{{"c", "Abc", std::nullopt, "Mountains and rivers"}}},
+      {{{}}},
+      {{{"Oceans and skies"}}},
+  });
+
+  testRoundTrip(data);
+}
```

Comment:
nit: Could we also have a test for nested complex vector (e.g., array of array or map of array)?

Reply:
That would be nice. Will try to do that in a follow-up.